bot/exts/info/doc/_cog.py

```diff
@@ -0,0 +1,423 @@
+
+            self.doc_symbols[overridden_symbol_name] = original_item
+            self.renamed_symbols[symbol_name].append(overridden_symbol_name)
+
+        # If we can't specially handle the symbol through its group or package,
+        # fall back to prepending its package name to the front.
+        else:
+            new_symbol_name = f"{package_name}.{symbol_name}"
+            if new_symbol_name in self.doc_symbols:
+                # If there's still a conflict, add the symbol's group in the middle.
+                new_symbol_name = f"{package_name}.{group_name}.{symbol_name}"
+            self.renamed_symbols[symbol_name].append(new_symbol_name)
+            return new_symbol_name
+
+    async def refresh_inventories(self) -> None:
+        """Refresh internal documentation inventories."""
+        self.refresh_event.clear()
+        await self.symbol_get_event.wait()
+        log.debug("Refreshing documentation inventory...")
+        self.inventory_scheduler.cancel_all()
+
+        self.base_urls.clear()
+        self.doc_symbols.clear()
+        self.renamed_symbols.clear()
+        await self.item_fetcher.clear()
+
+        coros = [
+            self.update_or_reschedule_inventory(
+                package["package"], package["base_url"], package["inventory_url"]
+            ) for package in await self.bot.api_client.get('bot/documentation-links')
+        ]
+        await asyncio.gather(*coros)
+        log.debug("Finished inventory refresh.")
+        self.refresh_event.set()
+
+    async def get_symbol_embed(self, symbol_name: str) -> Optional[discord.Embed]:
+        """
+        Attempt to scrape and fetch the data for the given `symbol_name`, and build an embed from its contents.
+
+        If the symbol is known, an Embed with documentation about it is returned.
+
+        First check the DocRedisCache before querying the cog's `BatchParser`.
+        """
+        log.trace(f"Building embed for symbol `{symbol_name}`")
+        if not self.refresh_event.is_set():
+            log.debug("Waiting for inventories to be refreshed before processing item.")
+            await self.refresh_event.wait()
+
+        doc_item = self.doc_symbols.get(symbol_name)
+        if doc_item is None:
+            if symbol_name.count(" "):
+                # If an invalid symbol contains a space, check if the command was invoked
+                # in the format !d <symbol> <message>
+                symbol_name = symbol_name.split(" ", maxsplit=1)[0]
+                doc_item = self.doc_symbols.get(symbol_name)
+                if doc_item is None:
+                    log.debug("Symbol does not exist.")
+                    return None
+            else:
+                log.debug("Symbol does not exist.")
+                return None
+
+        self.bot.stats.incr(f"doc_fetches.{doc_item.package}")
+
+        with self.symbol_get_event:
+            markdown = await doc_cache.get(doc_item)
```

Comment:
If the coro suspends here, and a refresh deleting the `doc_item` (or an equal value) occurs then the `get_markdown` would fail, thinking about it I'm not sure if it's possible depending on how asyncio works but I'd rather be safe unless you think it's unnecessary.
All the other code in the function is either synchronous or safe to switch on

Reply:
Refreshing inventories doesn't clear the cache.

Reply:
The containers in the `BatchParser` will be cleared and an old `doc_item` will be passed to `get_markdown` if it continues from here

Reply:
Using this event here doesn't prevent such situation: as soon as an item is fetched from redis (or not found), the event will be set and the refresh will finally be allowed to continue. It will end up clearing everything potentially before `get_markdown` is called. 

To avoid crashes in `get_markdown`, store the relevant future as a local variable rather than always accessing it through the dict.

Reply:
Actually, that last suggestion is a bad idea since the future will never set and it will wait forever.

Reply:
Can it refresh before `get_markdown` is called? The event will be set when the last get finishes and the context mgr is exited, after which the next context switch can only occur after `user_requested` is set to true on the item in `get_markdown` making the clear wait for that to be parsed

Reply:
Looks like you're right, but I believe allowing a refresh to run in the middle of `get_markdown` is still a problem. The two problems are that `_page_doc_items` and `_item_futures`, causing `KeyError`s later in `get_markdown`.

Reply:
Do you propose to just wrap it whole in the symbol_get_event? I suppose it would make it less likely for race conditions to occur in future changes even if it should be safe now

Reply:
I haven't looked over the new commits yet, but my point was that it wasn't safe. There was nothing to prevent a refresh from running during this code and I had outlined why that happening would be problematic.

Reply:
I've looked over everything now. Help me understand if this is correct:

`get_markdown` has two awaits (the get request and the executor) which are opportunities for a refresh to occur. If a refresh occurs, it clears out the futures. Once `get_markdown` resumes, it will still add the symbols to the queue. At the end, it will create a new future (thanks to `defaultdict`) and await it. Because the symbols were queued and the queue will still be consumed, the queue will eventually get the newly create future and set it regardless. The only problem is that `user_requested` will not be `True` for the future created.

Reply:
Unless I missed something, after the `get_markdown` coro is called, the `user_requested` attribute of the future is immediately set to `True` before a refresh can occur (currenty L102 and L118). 
Then if a refresh occurs at the two points you mentioned, it'll clear the cog's containers (which we don't care about as new doc gets are suspended until it finishes), afterwards it'll clear the `BatchParser` which will await every future where the `user_requested` attr is `True` before the task is cancelled and any of the containers are cleared. This prevents issues where they'd be cleared in the middle of getting an user item as the refresh won't run until there's nothing important left to parse and it won't receive new requests because the cog waits for `self.refresh_event`

Reply:
You're right. I overlooked the fact that the clear waits for the futures. The future will already exist by the type a refresh can be called, so it will be forced to wait for it.

> Unless I missed something, after the get_markdown coro is called, the user_requested attribute of the future is immediately set to True before a refresh can occur (currenty L102 and L118).

Yes. However, if a refresh was able to delete the future (it can't since it awaits it), then it'd force a new future to be created at the return statement.

Reply:
Well, it clearly is a bit confusing with the synchronization around it. I'll just move the body after the refresh event wait into the get_event and add a comment so the future changes are aware of its use without going over the awaits themselves